### PR TITLE
Notifications

### LINKED ISSUE #145: Let people pick when they want text messages
From Mike Meeting - not necessary, but cool feature.

### PR DESCRIPTION
**Pull request by bendelonlee**
Adds a settings controller, scss, and view. Adds settings to the nav bar.
Moves the telephone field from the dashboard to settings#edit.
Adds settings columns to the users table. 
Adds a user notifications spec file.
Adds javascript to the settings edit page.
Adds a rainy day mailer.
Does not add a rainy day texter. We should decide if we want to be doing texting.
Adds validations to the user model so that telephone must be present to have receives_texts be true

Adds a setting for frost notice, but no implementation. To implement we may want to add a temp_low column to database? That way we could find all the users and their frosty gardens with one query.  Unlike what we're doing for rainy day notice, which is not ideal (I wrote it, but it's not great).  Also, should weather data be on the zip code table so we don't have to store the same data multiple places? Given the scale of our app at this point, I don't see this as a pressing issue per se, and might be a pain to refactor, just a thought.

Other small changes:
Fixes the drag and drop bug we were having by detaching but not removing the parent div. (!!!!)
Makes the dark sky image into a link.
Makes a button_to with method: :get into a link_to to avoid quirks. It's already wrapped in button tags and visually looks exactly the same.

closes #145

### CODE REVIEW

File: app/assets/javascripts/schedule.js [hunk 1]
```diff
@@ -31,9 +31,9 @@ $( document ).ready(function() {
         garden = emptyParent
       }
       if (draggedFrom.find('.watering').length === 2) {
-        draggedFrom.remove();
+        draggedFrom.detach();
```
Comment (le3ah) [inline app/assets/javascripts/schedule.js]: Good catch!

File: app/controllers/settings_controller.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
+class SettingsController < ApplicationController
```
Comment (le3ah) [inline app/controllers/settings_controller.rb]: I like it!

File: app/mailers/rainy_day_mailer.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+class RainyDayMailer < ApplicationMailer
+  def inform(user, garden, chance)
+    @user = user
+    @garden = garden
+    @chance = chance
+    @attribution = user.own_gardens.include?(garden) ? "your garden" : "#{user.first_name}'s garden"
+    mail(to: user.email, subject: "Heads up!")
```
Comment (le3ah) [inline app/mailers/rainy_day_mailer.rb]: 🗣 🌱

File: app/models/user.rb [hunk 4]
```diff
@@ -35,4 +37,16 @@ def caretaking_gardens
...
+
+  def telephone_if_receives_texts
+    if (telephone.nil? || telephone.empty?) && receives_texts
+      errors.add(:phone_number, "can't be blank if you'd like to receive texts")
+    end
+  end
+
+  def telephone_ten_digits
+    if telephone && telephone.size > 0 && (telephone[/\d{10}/] != telephone)
+      errors.add(:phone_number, "must be ten digits")
```
Comment (le3ah) [inline app/models/user.rb]: Good call

File: app/views/settings/_notification_inputs.html.erb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  <div class="form-group ">
+    <%= f.check_box(:rainy_day_notifications) %>
+    <%= f.label(:rainy_day_notifications) %>
+    <p>We'll let you know every morning at 8 when rain is likely in the forecast so you can consider giving your plants a moisture break.</p>
+  </div>
+  <br>
+  <div class="form-group ">
+    <%= f.check_box(:frost_notifications) %>
+    <%= f.label(:frost_notifications) %>
+    <p>We'll let you know the night before when frost is likely in the forecast so you can take protective measures.</p>
```
Comment (le3ah) [inline app/views/settings/_notification_inputs.html.erb]: Hooray!

File: app/views/users/show.html.erb [hunk 6]
```diff
@@ -106,5 +104,7 @@
   </div>
 <% end %>
 <div class="darksky-container">
-  <%= image_tag(asset_path("icons/poweredby-darksky.png"), class: "darksky-img")  %>
+  <%= link_to("https://darksky.net/poweredby/" ) do %>
```
Comment (le3ah) [inline app/views/users/show.html.erb]: Great idea making this a live link

File: spec/features/users/user_receives_notifications_spec.rb [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    end
+    it 'only lets you receive texts if you give your correct phone number' do
+      visit settings_path
+      find(:css, "#user_receives_texts").set(true)
+      click_on("Save")
+      expect(page).to have_content("Phone number can't be blank if you'd like to receive texts")
+      expect(@user.reload.receives_texts).to eq(false)
+      fill_in :user_telephone, with: "789"
+      click_on("Save")
+      expect(@user.reload.receives_texts).to eq(false)
```
Comment (le3ah) [inline spec/features/users/user_receives_notifications_spec.rb]: Future sad path testing could be filling in with 11 or more digits or filling in with alphabetical characters.

File: spec/jobs/rainy_day_job_spec.rb [hunk 8]
```diff
@@ -35,30 +35,62 @@ def stub_twillio
...
 
+      expect(@mock_texter).to have_received(:send_rainy_day_text).with(user_1, shared_garden, 80.0)
+      expect(@mock_texter).to have_received(:send_rainy_day_text).with(caretaker, shared_garden, 80.0)
+      expect(@mock_texter).to have_received(:send_rainy_day_text).with(other_user, unshared_garden, 80.0)
+    end
+    it 'refrains from sending users texts, while sending email' do
+      unshared_garden = Garden.create!(@rainy_garden)
+      shared_garden = Garden.create!(@rainy_garden)
+      user_1 = create(:user, gardens: [shared_garden], telephone: ENV['ADMIN_PHONE_NUMBER'], receives_texts: false, receives_emails: true, rainy_day_notifications: true)
+      caretaker = create(:user, telephone: ENV['ADMIN_PHONE_NUMBER'], receives_texts: false, receives_emails: false, rainy_day_notifications: false)
```
Comment (le3ah) [inline spec/jobs/rainy_day_job_spec.rb]: I like this test

File: app/assets/javascripts/schedule.js (hunk 1 above)
Comment (asmolentzov) [reply app/assets/javascripts/schedule.js]: yay!

File: app/jobs/rainy_day_job.rb [hunk 9]
```diff
@@ -2,25 +2,45 @@ class RainyDayJob < ApplicationJob
...
 
   private
 
-  def text_users
-    @num_texts = 0
-    RainyDay.generate_rainy_days.each do |rainy_day|
-      send_rainy_day_texts(rainy_day)
+  def inform_users
+    @num_texts = 0; @num_emails = 0
+    RainyDay.generate_rainy_days.uniq(&:zip_code).each do |rainy_day|
```
Comment (asmolentzov) [inline app/jobs/rainy_day_job.rb]: nice!

File: app/mailers/rainy_day_mailer.rb [hunk 10]
```diff
@@ -0,0 +1,10 @@
+class RainyDayMailer < ApplicationMailer
+  def inform(user, garden, chance)
+    @user = user
+    @garden = garden
+    @chance = chance
+    @attribution = user.own_gardens.include?(garden) ? "your garden" : "#{user.first_name}'s garden"
```
Comment (asmolentzov) [inline app/mailers/rainy_day_mailer.rb]: this is nice!

File: app/models/user.rb [hunk 11]
```diff
@@ -1,6 +1,8 @@
 class User < ApplicationRecord
   has_many :user_gardens
   has_many :gardens, through: :user_gardens
+  validate :telephone_if_receives_texts
```
Comment (asmolentzov) [inline app/models/user.rb]: this is cool, I like how you're handling the errors with these

**Pull request merged**